Add how-to fork and rebase a fork (#370)
huggingface/huggingface_hub: docs/hub/model-repos.md

$$ E=mc^2 $$

### How can I fork a repository with LFS pointers ?

> When you want to fork a repository with LFS files, you cannot use the usual git way, and you need to be careful to not break the LFS pointers.

Reviewer: I think rebase and forking ended up being pretty much the same due to the initial commit automatically created for all repos. Only one line differs. Here is a suggestion of how to format this section, but feel free to decline it :)

### How can I fork or rebase a repository with LFS pointers?

When you want to fork or [rebase](https://git-scm.com/docs/git-rebase) a repository with [LFS](https://git-lfs.github.com/) files (all files over 20MB are stored as such), you cannot use the usual Git approach, and you need to be careful to not break the LFS pointers. Forking can take time depending on your bandwidth, because you will have to fetch an re-upload all the LFS files in your fork.

For example, say you have an upstream repository, **upstream**, and you just created your own repository on the Hub which is **myfork** in this example.

1. Create a destination repo (e.g. **myfork**) in https://huggingface.co/. 

2. Clone your fork repository

```
git lfs clone https://huggingface.co/me/myfork.git
```

3. Fetch non LFS files

```
cd myfork
git lfs install --skip-smudge --local # affects only this clone
git remote add upstream https://huggingface.co/friend/upstream.git
git fetch upstream
```

4. Fetch large files. This can take some time depending on your download bandwidth

```
git lfs fetch --all upstream # this can take time depending on your download bandwidth
```

4.a. If you want to override completely the fork history (which should only have an initial commit), run:

```
git reset --hard upstream/main
```

4.b. If you want to rebase instead of overriding, run the following command and solve any conflicts

```
git rebase upstream/main
```

5. Prepare your LFS files to push

```
git lfs install --force --local # this reinstalls the LFS hooks
huggingface-cli lfs-enable-largefiles . # needed if some files are bigger than 5Gb
```

6. And finally push

```
git push --force origin main # this can take time depending on your upload bandwidth
```

Now you have your own fork or rebased repo in the Hub!

Reviewer 2: > due to the initial commit automatically created for all repos

you can always `push --force` if you don't want to keep the initial commit (or we could maybe add an option to not create an initial commit at repo creation, like on GitHub)

Reviewer: > or we could maybe add an option to not create an initial commit at repo creation, like on GitHub

I would not do this (1) so we keep simplicity and (2) since `.gitattributes` has an important role in HF.

> you can always push --force if you don't want to keep the initial commit

If you think splitting the two use cases is very useful, let's do that, but I personally like that you just follow the same instructions except step 4

Reviewer 2: no what you propose looks great from a quick glance!

Author: this is great @osanseviero thanks

Reviewer: From user point of view it might be a bit annoying to having to copy each step, but it's also a way to educate them in the Git ways 🧙  (alternatively we can also have all the steps in a single snippet at the end)

Author: well, with long running commands, you better not copy paste the whole serial block anyways...

Author: I pushed your rework

$$ E=mc^2 $$

### How can I fork or rebase a repository with LFS pointers?

> When you want to fork or [rebase](https://git-scm.com/docs/git-rebase) a repository with [LFS](https://git-lfs.github.com/) files (all files over 20MB are stored as such), you cannot use the usual Git approach, and you need to be careful to not break the LFS pointers. Forking can take time depending on your bandwidth, because you will have to fetch an re-upload all the LFS files in your fork.

Reviewer: ```suggestion
When you want to fork or [rebase](https://git-scm.com/docs/git-rebase) a repository with [LFS](https://git-lfs.github.com/) files (all files over 20MB are stored as such), you cannot use the usual Git approach since you need to be careful to not break the LFS pointers. Forking can take time depending on your bandwidth, because you will have to fetch an re-upload all the LFS files in your fork.
```

For example, say you have an upstream repository, **upstream**, and you just created your own repository on the Hub which is **myfork** in this example.

> 1. Create a destination repo (e.g. **myfork**) in https://huggingface.co/. 

Reviewer: ```suggestion
1. Create a destination repository (e.g. **myfork**) in https://huggingface.co 
```

University of Illinois/NCSA Open Source License	| `ncsa`
The Unlicense	| `unlicense`
zLib License	| `zlib`

### How can I fork a repository with LFS pointers ?

> With "repoA" as the original repository forked in "repoB" as your repository, follow this procedure in order not to break LFS files while rebasing your fork "repoB" with "repoA". If you don't need to rebase, ignore the `git checkout -b temp repoA/main` and `rebase` part, just use the `git lfs *` parts.

Reviewer 3: I think something like this may sound a little clearer :) :

```suggestion
When you want to fork a repository with LFS pointers, you need to be careful you don’t break the LFS files when you rebase your fork.

For example, say you have an original repository, **repoA**, and it’s fork, **repoB**. Here is how you can safely fork and rebase **repoB** with **repoA** without breaking anything:
```

Reviewer 3: I would also consider placing this section above the list of licenses, since users will probably be more interested in reading this.

```
git lfs clone https://huggingface.co/me/repoB.git
cd repoB
git lfs install --skip-smudge --local # affects only this clone
git remote add repoA https://huggingface.co/friend/repoA.git
git fetch repoA

> git checkout -b temp repoA/main

Reviewer 3: And then we can add this piece here if you don't need to rebase:

 ```suggestion
git checkout -b temp repoA/main # ignore if you don't need to rebase
git rebase main # ignore if you don't need to rebase
```
